h2o-r/h2o-package/R/models.R

```diff
@@ -3189,6 +3192,56 @@ h2o.getModelTree <- function(model, tree_number, tree_class = NA) {
     tree@tree_class <- res$tree_class
   }
   
+  if(is.logical(res$levels)){ # Vector of NAs is recognized as logical type in R
+    tree@levels <- rep(list(NULL), length(res$levels))
+  } else {
+    tree@levels <- res$levels
+  }
+  
+  for (i in 1:length(tree@levels)){
+    if(!is.null(tree@levels[[i]])){
+    tree@levels[[i]] <- tree@levels[[i]] + 1
+    }
+  }
+  
+  # Convert numerical categorical levels to characters
+  pointer <-as.integer(1);
+  for(i in 1:length(tree@left_children)){
+    
+    right <- tree@right_children[i];
+    left <- tree@left_children[i]
+    split_column_cat_index <- match(tree@features[i], model@model$columns$names) # Indexof split column on children's parent node
+    if(is.na(split_column_cat_index)){ # If the split is not categorical, just increment & continue
+      if(right != -1) pointer <- pointer + 1;
+      if(left != -1) pointer <- pointer + 1;
+    }
+    
+    # Left child node's levels converted to characters
```

Comment:
This code is duplicated for left & right. Could be extracted, yet I wanted to avoid having another private method starting with .h2o ...

Reply:
There is no need to create a private function, in R you can define a function inside a function (to be only used in this function).

Reply:
Also: you technically only need to represent the variables that go LEFT, the RIGHT ones are complementary to the ones that go LEFT.

Reply:
This is correct. It can be done for the backend part (sending indices to the client).

However, regarding the client, I have a question. As the tree is walked, the predecessor to current node and the right/left sibling is of course known. It is therefore only required to know the subset of categorical levels of one node, as the other one is complementary, as you say. Question is, if we really want to do that, as it would make it harder to traverse the tree on the client for people who will not want to use our tree walker functionality.

I'd basically say you're right completely, because nobody will really abandon tree walker in favor of walking the three themselves, right ? And if so, it should be considered "expert mode". So I will start doing it like this, please only confirm my thoughts.

Reply:
It doesn't have to be harder to walk the tree - the user-facing structures can still look the same. My point is that the information doesn't have to be transferred to the client from the backend. The client can calculate the complement and expose it to the user.